File: Unity App/RIFT/Assets/Rift/RiftMdns.cs
// RiftMdns.cs - Zeroconf: publish this instance as _rift._tcp and watch for
// other RIFT instances and ComCentre (_flask-link._tcp) - browsing both is what
// lets DREAM show up in this dashboard without ComCentre knowing anything about
// RIFT. A small responder + browser over multicast UDP (224.0.0.251:5353) using
// just the DNS wire format: PTR (service -> instance), SRV (instance -> host +
// port), TXT and A (host -> address). No Bonjour/Avahi or extra package needed,
// so it behaves the same inside Unity on Windows, Linux and macOS.

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Rift
{
    public sealed class RiftMdns : IDisposable
    {
        const ushort TA = 1, TPtr = 12, TTxt = 16, TSrv = 33, TAny = 255;
        const ushort ClassIn = 1, CacheFlush = 0x8000;
        static readonly IPAddress Group = IPAddress.Parse("224.0.0.251");

        readonly RiftConfig cfg;
        readonly RiftPeers peers;
        Socket sock;
        volatile bool running;
        Thread recvThread, queryThread;

        // Only touched from the receive thread.
        readonly Dictionary<string, KeyValuePair<int, string>> srv = new Dictionary<string, KeyValuePair<int, string>>(); // lowercase instance -> (port, host)
        readonly Dictionary<string, string> addrs = new Dictionary<string, string>();                                     // lowercase host -> IPv4
        readonly Dictionary<string, string> instances = new Dictionary<string, string>();                                 // lowercase instance -> original case

        public RiftMdns(RiftConfig cfg, RiftPeers peers)
        {
            this.cfg = cfg;
            this.peers = peers;
        }

        // Returns null on success, else an error message.
        public string Start()
        {
            try
            {
                sock = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
                sock.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
                sock.Bind(new IPEndPoint(IPAddress.Any, 5353));
                var local = IPAddress.Parse(cfg.Ip);
                try { sock.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(Group, local)); }
                catch (SocketException) { sock.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(Group)); }
                sock.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, local.GetAddressBytes());
                sock.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastLoopback, true);
                sock.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastTimeToLive, 255);
                sock.ReceiveTimeout = 500;
            }
            catch (Exception e) when (e is SocketException || e is FormatException)
            {
                return e.Message;
            }
            running = true;
            recvThread = new Thread(ReceiveLoop) { IsBackground = true, Name = "rift-mdns-recv" };
            queryThread = new Thread(QueryLoop) { IsBackground = true, Name = "rift-mdns-query" };
            recvThread.Start();
            queryThread.Start();
            return null;
        }

        public void Dispose()
        {
            if (!running) return;
            running = false;
            Send(AnnouncePacket(0)); // goodbye
            recvThread?.Join(1500);
            queryThread?.Join(1500);
            try { sock.Close(); } catch (SocketException) { }
        }

        // ---- building ----

        static void U16(List<byte> b, int v) { b.Add((byte)(v >> 8)); b.Add((byte)v); }
        static void U32(List<byte> b, long v) { U16(b, (int)(v >> 16) & 0xffff); U16(b, (int)v & 0xffff); }

        static void Name(List<byte> b, string name)
        {
            foreach (var label in name.TrimEnd('.').Split('.'))
            {
                var bytes = Encoding.UTF8.GetBytes(label);
                b.Add((byte)bytes.Length);
                b.AddRange(bytes);
            }
            b.Add(0);
        }

        static void Record(List<byte> b, string name, int type, int cls, long ttl, List<byte> rdata)
        {
            Name(b, name);
            U16(b, type);
            U16(b, cls);
            U32(b, ttl);
            U16(b, rdata.Count);
            b.AddRange(rdata);
        }

        static List<byte> Header(int flags, int qd, int an, int ar)
        {
            var b = new List<byte>();
            U16(b, 0); U16(b, flags); U16(b, qd); U16(b, an); U16(b, 0); U16(b, ar);
            return b;
        }

        byte[] AnnouncePacket(long ttl)
        {
            string service = "_rift._tcp.local", instance = cfg.Name + "." + service, host = cfg.Name + ".local";
            var p = Header(0x8400, 0, 1, 3);

            var ptr = new List<byte>();
            Name(ptr, instance);
            Record(p, service, TPtr, ClassIn, ttl, ptr);

            var srvData = new List<byte>();
            U16(srvData, 0); U16(srvData, 0); U16(srvData, cfg.Port);
            Name(srvData, host);
            Record(p, instance, TSrv, CacheFlush | ClassIn, ttl, srvData);

            var txt = new List<byte>();
            var entry = Encoding.UTF8.GetBytes("role=fleet_manager");
            txt.Add((byte)entry.Length);
            txt.AddRange(entry);
            Record(p, instance, TTxt, CacheFlush | ClassIn, ttl, txt);

            var a = new List<byte>(IPAddress.Parse(cfg.Ip).GetAddressBytes());
            Record(p, host, TA, CacheFlush | ClassIn, ttl, a);
            return p.ToArray();
        }

        static byte[] QueryPacket(string name, int type)
        {
            var p = Header(0, 1, 0, 0);
            Name(p, name);
            U16(p, type);
            U16(p, ClassIn);
            return p.ToArray();
        }

        void Send(byte[] packet)
        {
            try { sock.SendTo(packet, new IPEndPoint(Group, 5353)); }
            catch (SocketException) { }
            catch (ObjectDisposedException) { }
        }

        // ---- parsing ----

        sealed class Rec
        {
            public string Name = "", Target = "", Addr = "";
            public int Type, Ttl, Port;
        }

        static int Rd16(byte[] b, int p)
        {
            if (p + 2 > b.Length) throw new IndexOutOfRangeException();
            return (b[p] << 8) | b[p + 1];
        }

        // Reads a (possibly compressed) name at p; advances p past it.
        static string ReadName(byte[] b, ref int p)
        {
            var labels = new List<string>();
            int pos = p;
            bool jumped = false;
            int hops = 0;
            while (true)
            {
                if (pos >= b.Length) throw new IndexOutOfRangeException();
                int len = b[pos];
                if (len == 0) { pos++; break; }
                if ((len & 0xc0) == 0xc0)
                {
                    if (pos + 1 >= b.Length || ++hops > 32) throw new IndexOutOfRangeException();
                    if (!jumped) p = pos + 2;
                    jumped = true;
                    pos = ((len & 0x3f) << 8) | b[pos + 1];
                }
                else
                {
                    if (pos + 1 + len > b.Length) throw new IndexOutOfRangeException();
                    labels.Add(Encoding.UTF8.GetString(b, pos + 1, len));
                    pos += 1 + len;
                }
            }
            if (!jumped) p = pos;
            return string.Join(".", labels);
        }

        static void Parse(byte[] b, List<KeyValuePair<string, int>> questions, List<Rec> recs)
        {
            int qd = Rd16(b, 4), total = Rd16(b, 6) + Rd16(b, 8) + Rd16(b, 10);
            int p = 12;
            for (int i = 0; i < qd; i++)
            {
                string n = ReadName(b, ref p);
                questions.Add(new KeyValuePair<string, int>(n, Rd16(b, p)));
                p += 4;
            }
            for (int i = 0; i < total; i++)
            {
                var r = new Rec { Name = ReadName(b, ref p) };
                r.Type = Rd16(b, p);
                r.Ttl = (Rd16(b, p + 4) << 16) | Rd16(b, p + 6);
                int rdlen = Rd16(b, p + 8), rd = p + 10;
                if (rd + rdlen > b.Length) throw new IndexOutOfRangeException();
                if (r.Type == TPtr) { int q = rd; r.Target = ReadName(b, ref q); }
                else if (r.Type == TSrv) { r.Port = Rd16(b, rd + 4); int q = rd + 6; r.Target = ReadName(b, ref q); }
                else if (r.Type == TA && rdlen == 4) r.Addr = b[rd] + "." + b[rd + 1] + "." + b[rd + 2] + "." + b[rd + 3];
                recs.Add(r);
                p = rd + rdlen;
            }
        }

        static bool IsBrowsedService(string lname) => lname == "_rift._tcp.local" || lname == "_flask-link._tcp.local";
        static bool IsBrowsedInstance(string lname) => lname.Contains("._rift._tcp.local") || lname.Contains("._flask-link._tcp.local");

        void PublishPeers()
        {
            foreach (var kv in srv)
            {
                string instance = instances[kv.Key];
                string shortName = instance.Split('.')[0];
                if (shortName == cfg.Name) continue;
                if (!addrs.TryGetValue(kv.Value.Value.ToLowerInvariant(), out var ip)) continue;
                peers.Set(shortName, "http://" + ip + ":" + kv.Value.Key);
            }
        }

        void HandlePacket(byte[] packet)
        {
            var questions = new List<KeyValuePair<string, int>>();
            var recs = new List<Rec>();
            Parse(packet, questions, recs);

            string service = "_rift._tcp.local", instance = (cfg.Name + "." + service).ToLowerInvariant(), host = (cfg.Name + ".local").ToLowerInvariant();
            foreach (var q in questions)
            {
                string n = q.Key.ToLowerInvariant();
                int t = q.Value;
                if ((n == service && (t == TPtr || t == TAny)) || (n == instance && (t == TSrv || t == TTxt || t == TAny)) || (n == host && (t == TA || t == TAny)))
                {
                    Send(AnnouncePacket(120)); // someone is asking about us
                    break;
                }
            }

            foreach (var r in recs)
            {
                string lname = r.Name.ToLowerInvariant();
                if (r.Type == TPtr && IsBrowsedService(lname))
                {
                    string shortName = r.Target.Split('.')[0];
                    if (r.Ttl == 0) peers.Remove(shortName); // goodbye
                    else if (shortName != cfg.Name)
                    {
                        instances[r.Target.ToLowerInvariant()] = r.Target;
                        if (!srv.ContainsKey(r.Target.ToLowerInvariant())) Send(QueryPacket(r.Target, TSrv));
                    }
                }
                else if (r.Type == TSrv && IsBrowsedInstance(lname))
                {
                    instances[lname] = r.Name;
                    srv[lname] = new KeyValuePair<int, string>(r.Port, r.Target);
                    if (!addrs.ContainsKey(r.Target.ToLowerInvariant())) Send(QueryPacket(r.Target, TA));
                }
                else if (r.Type == TA && r.Addr.Length > 0) addrs[lname] = r.Addr;
            }
            PublishPeers();
        }

        void ReceiveLoop()
        {
            var buf = new byte[9000];
            while (running)
            {
                int n;
                try { n = sock.Receive(buf); }
                catch (SocketException) { continue; }        // timeout: re-check running
                catch (ObjectDisposedException) { break; }
                if (n <= 0) continue;
                try
                {
                    var packet = new byte[n];
                    Array.Copy(buf, packet, n);
                    HandlePacket(packet);
                }
                catch (Exception)
                {
                    // A malformed packet from someone else must never take the responder down.
                }
            }
        }

        void QueryLoop()
        {
            int announced = 0;
            while (running)
            {
                Send(QueryPacket("_rift._tcp.local", TPtr));
                Send(QueryPacket("_flask-link._tcp.local", TPtr));
                if (announced < 3) { Send(AnnouncePacket(120)); announced++; }
                for (int i = 0; i < (announced < 3 ? 10 : 100) && running; i++) Thread.Sleep(100);
            }
        }
    }
}

File: Android App/RiftMdns.kt
// RiftMdns.kt - Zeroconf: publish this instance as _rift._tcp and watch for
// other RIFT instances and ComCentre (_flask-link._tcp) - browsing both is what
// lets DREAM show up in this dashboard without ComCentre knowing anything about
// RIFT. A small responder + browser over multicast UDP (224.0.0.251:5353) using
// just the DNS wire format: PTR (service -> instance), SRV (instance -> host +
// port), TXT and A (host -> address). Written on java.net.MulticastSocket rather
// than Android's NsdManager so the same code runs (and is testable) on a JVM;
// on a phone the Activity holds a WifiManager.MulticastLock so the radio
// actually delivers the packets.
package com.example.riftcontrol

import java.io.ByteArrayOutputStream
import java.net.DatagramPacket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.MulticastSocket
import java.net.NetworkInterface
import java.net.SocketTimeoutException

class RiftMdns(private val cfg: RiftConfig, private val peers: RiftPeers, private val log: (String) -> Unit = {}) {
    private companion object {
        const val T_A = 1; const val T_PTR = 12; const val T_TXT = 16; const val T_SRV = 33; const val T_ANY = 255
        const val CLASS_IN = 1; const val CACHE_FLUSH = 0x8000
        val GROUP: InetAddress = InetAddress.getByName("224.0.0.251")
    }

    private var socket: MulticastSocket? = null
    @Volatile private var running = false
    private var recvThread: Thread? = null
    private var queryThread: Thread? = null

    // Only touched from the receive thread.
    private val srv = HashMap<String, Pair<Int, String>>()   // lowercase instance -> (port, target host)
    private val addrs = HashMap<String, String>()            // lowercase host -> IPv4
    private val instances = HashMap<String, String>()        // lowercase instance -> original case

    /** Returns null on success, else an error message. */
    fun start(): String? {
        try {
            val s = MulticastSocket(null)
            s.reuseAddress = true
            s.bind(InetSocketAddress(5353))
            val local = try { NetworkInterface.getByInetAddress(InetAddress.getByName(cfg.ip)) } catch (_: Exception) { null }
            if (local != null) {
                s.joinGroup(InetSocketAddress(GROUP, 5353), local)
                s.networkInterface = local
            } else {
                @Suppress("DEPRECATION") s.joinGroup(GROUP)
            }
            s.timeToLive = 255
            s.soTimeout = 500
            socket = s
        } catch (e: Exception) {
            return e.message ?: e.toString()
        }
        running = true
        recvThread = Thread({ receiveLoop() }, "rift-mdns-recv").apply { isDaemon = true; start() }
        queryThread = Thread({ queryLoop() }, "rift-mdns-query").apply { isDaemon = true; start() }
        return null
    }

    fun stop() {
        if (!running) return
        running = false
        send(announcePacket(0)) // goodbye
        recvThread?.join(1500)
        queryThread?.join(1500)
        socket?.close()
    }

    // ---- building ----

    private fun ByteArrayOutputStream.u16(v: Int) { write(v shr 8 and 0xff); write(v and 0xff) }
    private fun ByteArrayOutputStream.u32(v: Long) { u16((v shr 16 and 0xffff).toInt()); u16((v and 0xffff).toInt()) }

    private fun ByteArrayOutputStream.name(n: String) {
        for (label in n.trimEnd('.').split('.')) {
            val bytes = label.toByteArray(Charsets.UTF_8)
            write(bytes.size)
            write(bytes)
        }
        write(0)
    }

    private fun ByteArrayOutputStream.record(n: String, type: Int, cls: Int, ttl: Long, rdata: ByteArray) {
        name(n); u16(type); u16(cls); u32(ttl); u16(rdata.size); write(rdata)
    }

    private fun header(flags: Int, qd: Int, an: Int, ar: Int) = ByteArrayOutputStream().apply {
        u16(0); u16(flags); u16(qd); u16(an); u16(0); u16(ar)
    }

    private fun announcePacket(ttl: Long): ByteArray {
        val service = "_rift._tcp.local"
        val instance = "${cfg.name}.$service"
        val host = "${cfg.name}.local"
        val p = header(0x8400, 0, 1, 3)
        p.record(service, T_PTR, CLASS_IN, ttl, ByteArrayOutputStream().apply { name(instance) }.toByteArray())
        p.record(instance, T_SRV, CACHE_FLUSH or CLASS_IN, ttl,
            ByteArrayOutputStream().apply { u16(0); u16(0); u16(cfg.port); name(host) }.toByteArray())
        val txt = "role=fleet_manager".toByteArray()
        p.record(instance, T_TXT, CACHE_FLUSH or CLASS_IN, ttl, byteArrayOf(txt.size.toByte()) + txt)
        p.record(host, T_A, CACHE_FLUSH or CLASS_IN, ttl, InetAddress.getByName(cfg.ip).address)
        return p.toByteArray()
    }

    private fun queryPacket(name: String, type: Int) = header(0, 1, 0, 0).apply { name(name); u16(type); u16(CLASS_IN) }.toByteArray()

    private fun send(packet: ByteArray) {
        try { socket?.send(DatagramPacket(packet, packet.size, GROUP, 5353)) } catch (_: Exception) { }
    }

    // ---- parsing ----

    private class Rec(val name: String, val type: Int, val ttl: Int, val target: String, val addr: String, val port: Int)

    private fun rd16(b: ByteArray, p: Int): Int {
        if (p + 2 > b.size) throw IndexOutOfBoundsException()
        return (b[p].toInt() and 0xff shl 8) or (b[p + 1].toInt() and 0xff)
    }

    /** Reads a (possibly compressed) name at [start]; returns the name and the position just past it. */
    private fun readName(b: ByteArray, start: Int): Pair<String, Int> {
        val labels = ArrayList<String>()
        var pos = start
        var next = -1
        var hops = 0
        while (true) {
            if (pos >= b.size) throw IndexOutOfBoundsException()
            val len = b[pos].toInt() and 0xff
            if (len == 0) { pos++; break }
            if (len and 0xc0 == 0xc0) {
                if (pos + 1 >= b.size || ++hops > 32) throw IndexOutOfBoundsException()
                if (next < 0) next = pos + 2
                pos = (len and 0x3f shl 8) or (b[pos + 1].toInt() and 0xff)
            } else {
                if (pos + 1 + len > b.size) throw IndexOutOfBoundsException()
                labels.add(String(b, pos + 1, len, Charsets.UTF_8))
                pos += 1 + len
            }
        }
        return Pair(labels.joinToString("."), if (next >= 0) next else pos)
    }

    private fun parse(b: ByteArray): Pair<List<Pair<String, Int>>, List<Rec>> {
        val qd = rd16(b, 4)
        val total = rd16(b, 6) + rd16(b, 8) + rd16(b, 10)
        var p = 12
        val questions = ArrayList<Pair<String, Int>>()
        repeat(qd) {
            val (n, after) = readName(b, p)
            questions.add(Pair(n, rd16(b, after)))
            p = after + 4
        }
        val recs = ArrayList<Rec>()
        repeat(total) {
            val (n, after) = readName(b, p)
            val type = rd16(b, after)
            val ttl = (rd16(b, after + 4) shl 16) or rd16(b, after + 6)
            val rdlen = rd16(b, after + 8)
            val rd = after + 10
            if (rd + rdlen > b.size) throw IndexOutOfBoundsException()
            var target = ""; var addr = ""; var port = 0
            when {
                type == T_PTR -> target = readName(b, rd).first
                type == T_SRV -> { port = rd16(b, rd + 4); target = readName(b, rd + 6).first }
                type == T_A && rdlen == 4 -> addr = (0 until 4).joinToString(".") { (b[rd + it].toInt() and 0xff).toString() }
            }
            recs.add(Rec(n, type, ttl, target, addr, port))
            p = rd + rdlen
        }
        return Pair(questions, recs)
    }

    private fun isBrowsedService(l: String) = l == "_rift._tcp.local" || l == "_flask-link._tcp.local"
    private fun isBrowsedInstance(l: String) = l.contains("._rift._tcp.local") || l.contains("._flask-link._tcp.local")

    private fun publishPeers() {
        for ((key, portAndHost) in srv) {
            val instance = instances[key] ?: continue
            val short = instance.substringBefore('.')
            if (short == cfg.name) continue
            val ip = addrs[portAndHost.second.lowercase()] ?: continue
            peers.set(short, "http://$ip:${portAndHost.first}")
        }
    }

    private fun handle(packet: ByteArray) {
        val (questions, recs) = parse(packet)
        val service = "_rift._tcp.local"
        val instance = "${cfg.name}.$service".lowercase()
        val host = "${cfg.name}.local".lowercase()
        for ((qname, t) in questions) {
            val n = qname.lowercase()
            if ((n == service && (t == T_PTR || t == T_ANY)) || (n == instance && (t == T_SRV || t == T_TXT || t == T_ANY)) ||
                (n == host && (t == T_A || t == T_ANY))) {
                send(announcePacket(120)) // someone is asking about us
                break
            }
        }
        for (r in recs) {
            val lname = r.name.lowercase()
            when {
                r.type == T_PTR && isBrowsedService(lname) -> {
                    val short = r.target.substringBefore('.')
                    if (r.ttl == 0) peers.remove(short) // goodbye
                    else if (short != cfg.name) {
                        instances[r.target.lowercase()] = r.target
                        if (!srv.containsKey(r.target.lowercase())) send(queryPacket(r.target, T_SRV))
                    }
                }
                r.type == T_SRV && isBrowsedInstance(lname) -> {
                    instances[lname] = r.name
                    srv[lname] = Pair(r.port, r.target)
                    if (!addrs.containsKey(r.target.lowercase())) send(queryPacket(r.target, T_A))
                }
                r.type == T_A && r.addr.isNotEmpty() -> addrs[lname] = r.addr
            }
        }
        publishPeers()
    }

    private fun receiveLoop() {
        val buf = ByteArray(9000)
        while (running) {
            try {
                val dp = DatagramPacket(buf, buf.size)
                socket?.receive(dp) ?: break
                handle(dp.data.copyOf(dp.length))
            } catch (_: SocketTimeoutException) {
                // re-check running
            } catch (e: java.io.IOException) {
                if (!running) break
            } catch (_: Exception) {
                // A malformed packet from someone else must never take the responder down.
            }
        }
    }

    private fun queryLoop() {
        var announced = 0
        while (running) {
            send(queryPacket("_rift._tcp.local", T_PTR))
            send(queryPacket("_flask-link._tcp.local", T_PTR))
            if (announced < 3) { send(announcePacket(120)); announced++ }
            var i = 0
            while (i++ < (if (announced < 3) 10 else 100) && running) Thread.sleep(100)
        }
    }
}
